paho/message_ids.go

```diff
@@ -56,16 +56,19 @@ type MIDs struct {
 func (m *MIDs) Request(c *CPContext) (uint16, error) {
 	m.Lock()
 	defer m.Unlock()
-	for i := uint16(1); i < midMax; i++ {
-		v := (m.lastMid + i) % midMax
-		if v == 0 {
-			continue
-		}
-		if inuse := m.index[v]; inuse == nil {
-			m.index[v] = c
-			m.lastMid = v
-			return v, nil
+	for i := uint16(0); i < midMax; i++ {
```

Comment:
Please don't feel obliged to accept this suggestion. I'm mainly adding it because I found the current version a little hard to understand (some of that is the comment, some relates to uint16 overflows...).

Might I suggest:

```
	for i := uint16(0); i < midMax; i++ {
		v := (m.lastMid+i)%midMax + 1
		if m.index[v-1] != nil {
			continue
		}
		m.index[v-1] = c
		m.lastMid = v
		return v, nil
	}
	// It may be possible to reuse lastMid
	if m.lastMid != 0 && m.index[m.lastMid-1] == nil {
		m.index[m.lastMid-1] = c
		return m.lastMid, nil
	}

	return 0, ErrorMidsExhausted
```

or, even better in my opinion (because it's very obvious what it's doing and avoids the need to take a remainder):

```
func (m *MIDs) Request(c *CPContext) (uint16, error) {
	m.Lock()
	defer m.Unlock()
	for i := m.lastMid + 1; i != 0; i++ { // scan from lastMid + 1 to end of range
		if m.index[i-1] != nil {
			continue
		}
		m.index[i-1] = c
		m.lastMid = i
		return i, nil
	}
	for i := uint16(0); i < m.lastMid; i++ { // scan from start of range to lastMid (inclusive)
		if m.index[i] != nil {
			continue
		}
		m.index[i] = c
		m.lastMid = i + 1
		return i + 1, nil
	}

	return 0, ErrorMidsExhausted
}
```

Reply:
Thanks for your review !! Your second opinion is easy to read, so I reflected it to my implemention with a few change.